### PR TITLE
Replace compex loops with index access, reuse pack_strings() and unpack_strings()

### PR DESCRIPTION
**Pull request by Wovchena**

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -12,3 +12,4 @@ endif()
 
 add_subdirectory(llm/cpp)
 target_compile_definitions(llm PRIVATE USER_OV_EXTENSIONS_PATH=\"$<TARGET_FILE:user_ov_extensions>\")
+target_link_libraries(llm PRIVATE user_ov_extensions)
```
Comment (ilya-lavrenov) [inline CMakeLists.txt:15]:
```suggestion
add_dependencies(llm user_ov_extensions)
```

File: llm/cpp/llm.cpp [hunk 2]
```diff
@@ -2,41 +2,24 @@
 // SPDX-License-Identifier: Apache-2.0
 
 #include <openvino/openvino.hpp>
+#include <utils.hpp>
```
Comment (ilya-lavrenov) [inline llm/cpp/llm.cpp:5]:
```suggestion
#include "utils.hpp"
```

Comment (yury-gorbachev) [inline llm/cpp/llm.cpp:5]: File name is actually ambiguous, not sure if this is not error prone. 
I understand that it is tokenizers design though...
cc: @apaniukov

File: llm/cpp/llm.cpp [hunk 3]
```diff
@@ -50,37 +33,28 @@ int main(int argc, char* argv[]) try {
...
-            }
-        }
+    std::vector<ov::Output<ov::Node>> inputs = model->inputs();
+    for (size_t idx = 2; idx < inputs.size(); ++idx) {
+        ov::PartialShape shape = inputs.at(idx).get_partial_shape();
+        shape[0] = BATCH_SIZE;
+        shapes.emplace(idx, shape);
     }
     model->reshape(shapes);
     ov::preprocess::PrePostProcessor p3(model);
```
Comment (yury-gorbachev) [inline llm/cpp/llm.cpp:51]: Why is this needed? What is original type in IR?

Comment (Wovchena) [reply llm/cpp/llm.cpp:51]: ov::element::i64

Comment (slyalin) [reply llm/cpp/llm.cpp:51]: Can we explicitly override the input element type with `i32` in intel-optimum code export? In the code below we always override the type, and actually it will likely be the case for any generative model. We have knowledge about the model in the export code and can do it safely instead of just blindly overriding it here in the code.

From the other hand, if i64 is not avoidable as an input type for the model, then why tokenizer gives i32 and requires alignment in the user code? Both sides are under our (partial) control. Let's make them composable from the beginning.

File: llm/cpp/llm.cpp (hunk 2 above)
Comment (slyalin) [reply llm/cpp/llm.cpp:5]: Add necessary definitions to `CMakeLists.txt` to include it as `tokenizer/utils.hpp` as `utils.hpp` is in that directory.

File: llm/cpp/llm.cpp [hunk 4]
```diff
@@ -50,37 +33,28 @@ int main(int argc, char* argv[]) try {
...
-                ov::PartialShape shape = input.get_partial_shape();
-                shape[0] = BATCH_SIZE;
-                shapes.emplace(name, shape);
-                break;
-            }
-        }
+    std::vector<ov::Output<ov::Node>> inputs = model->inputs();
+    for (size_t idx = 2; idx < inputs.size(); ++idx) {
+        ov::PartialShape shape = inputs.at(idx).get_partial_shape();
+        shape[0] = BATCH_SIZE;
```
Comment (slyalin) [inline llm/cpp/llm.cpp:47]: Set batch for all inputs in one loop, so start with `idx = 0` and remove initializer for `shapes` above. Dimension 0 is affected only in all inputs, no need to handle `input_ids` and `attention_mask` separately.

To better utilize OV capabilities, we can built-in layouts to the original model in optimum-intel export, and use `set_batch` function here here which will affect all inputs automatically. https://docs.openvino.ai/2023.1/openvino_docs_OV_UG_ShapeInference.html#the-set-batch-method.

File: llm/cpp/llm.cpp (hunk 3 above)
Comment (slyalin) [reply llm/cpp/llm.cpp:51]: We have `model_connect` or similarly called function that can connect tokenizer and main model. Here we don't need the connection itself, but the connection supposes similar alignment on types. Can we provide a way to align types as a service from the tokenizer library, @apaniukov? But it is better to already produce tokenizer model aligned anyway.

File: llm/cpp/llm.cpp (hunk 4 above)
Comment (Wovchena) [reply llm/cpp/llm.cpp:47]: I need to set channel dim to -1 because one of the models is exported by the notebook has static value here. So I can't remove shapes initializer

Comment (slyalin) [reply llm/cpp/llm.cpp:47]: Can we treat it a model issue? In case if we always expect incorrectly set shapes in loaded models, we always need to set all the dimensions. Does it really make sense if we control both parties?

Comment (Wovchena) [reply llm/cpp/llm.cpp:47]: I don't know. @eaidova, can I treat this as the notebook issue and ignore it in my implementation?

File: llm/cpp/llm.cpp (hunk 3 above)
Comment (AlexKoff88) [reply llm/cpp/llm.cpp:51]: > Can we explicitly override the input element type with `i32` in intel-optimum export code? In the code below we always override the type, and actually it will likely be the case for any generative model. We have knowledge about the model in the export code and can do it safely instead of just blindly overriding it here in the code.
> 
> From the other hand, if i64 is not avoidable as an input type for the model, then why tokenizer gives i32 and requires alignment in the user code? Both sides are under our (partial) control. Let's make them composable from the beginning.
> 
> @eaidova, @AlexKoff88, could you provide your feedback.

i32 works for most of the models, AFAIK. But we cannot guarantee that it works for all of them. I would keep i64 in the HF code. But I don't mind if we use i32 in this specific sample which is for a limited set of models.

Comment (apaniukov) [reply llm/cpp/llm.cpp:51]: Change the default tokenizer output type to i64. Output type can be declared in `convert_tokenizer` now.

Comment (Wovchena) [reply llm/cpp/llm.cpp:51]: I think detokenizer input should be changed the same way

File: llm/cpp/llm.cpp (hunk 2 above)
Comment (Wovchena) [reply llm/cpp/llm.cpp]: Added

File: llm/cpp/llm.cpp [hunk 5]
```diff
@@ -50,61 +33,50 @@ int main(int argc, char* argv[]) try {
...
-                ireq.get_tensor(input).set_shape(input.get_partial_shape().get_min_shape());
-                break;
-            }
-        }
+    ov::InferRequest ireq = core.compile_model(model, "CPU", ov::cache_dir("llm-cache")).create_infer_request();
+    for (size_t idx = 2; idx < inputs.size(); ++idx) {
+        ireq.get_input_tensor(idx).set_shape(inputs.at(idx).get_partial_shape().get_min_shape());
     }
     ireq.get_tensor("input_ids").set_shape(input_ids.get_shape());  // TODO: replace with ireq.set_tensor("input_ids", input_ids); after it's fixed
     ireq.get_tensor("attention_mask").set_shape(input_ids.get_shape());
```
Comment (ilya-lavrenov) [inline llm/cpp/llm.cpp:59]: should we use `attention_mask` instead of `input_ids` to configure `"attention_mask"` input?

Comment (Wovchena) [reply llm/cpp/llm.cpp:59]: It doesn't matter, their shapes are identical. This code shouldn't have existed any way if there were no bugs in openvino

File: llm/cpp/llm.cpp (hunk 3 above)
Comment (apaniukov) [reply llm/cpp/llm.cpp]: @Wovchena yes, you are right. Pushed this change to the branch.

Comment (Wovchena) [reply llm/cpp/llm.cpp]: Updated

**Review by slyalin — APPROVED** [review]
Let's merge this part to move forward.
